### ISSUE TITLE
memory leak issues when running pymc3 model in a loop

### ISSUE THREAD

**Issue by mydatech001** [pymc-devs/pymc#1825]
I came across a memory leak problem. Any advice would be highly appreciated! 

1. In my experiment (win7, python3.5, pymc3.1), I need to run the same Bayesian linear model against multiple data sets. 
2. The model structure is the same and I've coded as a loop, i.e.  within each loop
2.1. specify the model structure first and then 
2.2  apply the same model against a new data set
3) The problem is the memory consumption keeps going up (10MB per iteration) even though i've used set (var=None and del Var ). I have got total 1000 iteration to go and the program stop after 300 iteration due to memory running-out problem (i tried memoryprofiler and it seems the memory consumed by pm.sample is not released)
4) This problem is very similar to the issue listed here [#772: "Final issue, is that running the code above in a loop seems to give a huge memory leak" ](https://github.com/pymc-devs/pymc3/issues/772) but i've seen no clear solution in the follow-up comments
5) Ideally, I'd like to specify the model structure once outside the loop and then apply the same model against multiple data sets in each iteration of the for_loop (similar to RJAGS way of coding)

The example codes copied below could repeat the problem as observed in my code (i.e. with the for_loop running, over the time, the memory consumption will keep building up till my computer running out of memory)

`

    import gc
    import numpy as np
    import matplotlib.pyplot as plt
    # Initialize random number generator
    np.random.seed(123)
    # True parameter values
    alpha, sigma = 1, 1
    beta = [1, 2.5]
    # Size of dataset
    size = 10000
    # Predictor variable
    X1 = np.random.randn(size)
    X2 = np.random.randn(size) * 0.2
    # Simulate outcome variable
    Y = alpha + beta[0]*X1 + beta[1]*X2 + np.random.randn(size)*sigma
    
    #%matplotlib inline
    fig, axes = plt.subplots(1, 2, sharex=True, figsize=(10,4))
    axes[0].scatter(X1, Y)
    axes[1].scatter(X2, Y)
    axes[0].set_ylabel('Y'); axes[0].set_xlabel('X1'); axes[1].set_xlabel('X2');
    
    from pymc3 import Model, Normal, HalfNormal
    
    for iteration in range (0,300):
        print(iteration)
        gc.collect()
    
        basic_model = Model()
        with basic_model:
            # Priors for unknown model parameters
            alpha = Normal('alpha', mu=0, sd=10)
            beta = Normal('beta', mu=0, sd=10, shape=2)
            sigma = HalfNormal('sigma', sd=1)
            # Expected value of outcome
            mu = alpha + beta[0]*X1 + beta[1]*X2
            # Likelihood (sampling distribution) of observations
            Y_obs = Normal('Y_obs', mu=mu, sd=sigma, observed=Y)
            
        from pymc3 import NUTS, sample
        with basic_model:
            # draw 2000 posterior samples
            trace = sample(5000, n_init=20000)

`

**Comment by hvasbath** [comment]
Yes I observed the same thing, running multiple chains in parallel. Simply do not return the trace in the last line. 
Just do:
sample(5000, n_init=20000)

And load the traces afterwards again using the pm.backend.load method when you need the results. Of course this is not the proper solution, but at least for me it was a solution ;) . I hope it works for you as well!

edit: Ah yes of course you are going to need to use a backend that stores the traces to the disc and not the NdArray backend. There is either Text or sqlite, or I saw hdf5 floating around recently, maybe that is merged as well already.

edit2: And you wont need to rebuild the model in each iteration- as you do not change anything in there. Just put the loop around the sample block.

**Comment by aaronjmaxwell** [comment]
Hello, I have a similar problem.  I have a rather large data set that I'm trying to fit a two parameter model to.  In order to prevent over fitting, I'm use scikit-learn KFold to fold my pandas DataFrame several times.  I fit the same Bayesian model (similar to what was shown above) but I only want the MAP.  However, every execution of find_MAP increases the memory usage of the entire program.  My use of the pyMC3 module is in a separate function call, like:

```
def BayesModel(df):
    import pymc3
    mymodel = pymc3.Model()
    with mymodel:
        x = pymc3.Normal(*args, **kwargs)
        ....
        posterior = pymc3.Bernoulli(*args, **kwargs)
    MAP = find_MAP(model = mymodel)
    # place MAP into DataFrames
    return DataFrames

import pandas
import numpy
df = pandas.read_csv('mydata.txt')
for i, j in Fold(df)
    vars = BayesModel(folded_df)
```
However, the memory jumps by a few hundred MB on a relatively small model (2000 vars, 500 MB data set).  Using the default minimizer results in total memory usage of 1731.6, 2391.9, 2911.9, 3479.2, 4086.7 MB.  Similar jumps are seen using the L-BFGS-B optimizer in scipy.optimize.

**Comment by twiecki** [comment]
@nouiz Could this be a theano problem? Does theano somehow retain existing graphs in RAM?

**Comment by mydatech001** [comment]
Thank you hvasbath!

1.） I've tried but failed to observe any improvement in my machine. i'm wondering what's the computer setting in your case. in my setting, it's a Win7 with Anaconda Python 3.5 and latest development version Pymc3(3.1) and Theano(dev9.0). I observe the same memory leak issue in my Linux machine as well as PYMC2  based solution)
`

    with basic_model:
        db = backends.SQLite('db.sqlite')
        sample(50000, n_init=20000, trace=db)

`
2） regarding your 2nd advice for looping sampling part only, i'm afraid this is not an option. I've tried it but the computational results would become incorrect if i apply the old model to the new data (to get correct results, i have to create a new instance of the model for each new data set even thought the model structure was exactly the same). i don't have the knowledge to understand why so but in my wild guess, the PYMC3 doesn't decouple the data instance with the model structure specification in its implementation..

**Comment by hvasbath** [comment]
I am using Ubuntu 12.04 on one machine Ubuntu14.04 on another machine. Each python 2.7 only with both the bleeding edges pymc3 and theano.

Regarding your looping over the model I cannot see where your input data changes- did you post the full script?
Still you can have the looping over only the sampling as you can simply update your observations in the course of the loop. Make your data shared variable and update that every iteration in the loop.

```
shared_data = theano.shared(data, borrow=True)

with pm.Model() as model:
    # model here ...
    Y_obs = Normal('Y_obs', mu=mu, sd=sigma, observed=shared_data)

for iteration in range(0, 300):
    #where ever your data is being changed as I said I cant see it
    shared_data.set_value(new_data)

    with model:
        sample(whatever ...)
```

I guess in your case the memory leak is then really created by rebuilding the models ergo also the graphs. That means you also have n-times compiled theano functions in memory each having all the contents of their respective graphs attached, input data etc. ... So this is not really a leak as this is the natural behaviour of things- its a feature.

It is never necessary to rebuild any model graph unless the way of computation changes. 
If the input data changes as mentioned above simply update your shared variables, which you anyway have to learn to use, because it is really inefficient and bug prone if you dont.

So no, pymc3 keeps the data seperate from the model. Although, of course at some point they have to interact ;) .

I remember other people were asking for a tutorial or example at least a statement about shared variables? Is this somewhere in the docs @twiecki ? 

Good luck getting it running @mydatech001 !

**Comment by mydatech001** [comment]
Super!!! The problem is solved by using the 'shared_data'. Thank you hvasbath. I'm a bit surprised why this important 'trick' is not shown in the PYMC3 tutorial on the web (my original purpose is to learn how to use PYMC3 and hence didn't pay any attention to Theano library).

**Comment by hvasbath** [comment]
Yeah, it is not really a pymc3 thing, it is basic Theano handling, but I guess it really should be mentioned in the tutorial anyways. I am glad it works now ;) . So please close the issue ;) .

**Comment by aaronjmaxwell** [comment]
Does this not apply if the shape of the model distribution itself changes?  Say I have a Pandas DataFrame X, and I take some fold / fraction of the data to fit my model:

```
X = pandas.DataFrame()
Fold = sklearn.model_selection.KFold(n_splits = n_splits):
for train_idx, test_idx in Fold.split(df):
    uniq_v1 = X.iloc[train_idx]['var1'].unique()
    shape_v1 = len(uniq_v1)
    uniq_v2 = X.iloc[train_idx]['var2'].unique()
    shape_v2 = len(uniq_v2)
    ...
    obs = X.iloc[train_idx]['observation']
    shape_obs = len(train_idx)
```
So I need to fit the same "type" of model, but the "shape" argument for each PyMC3 distribution will change depending on the fold.  Originally, I had put this model in the loop:
```
X = pandas.DataFrame()
Fold = sklearn.model_selection.KFold(n_splits = n_splits):
for train_idx, test_idx in Fold.split(df):
    uniq_v1 = X.iloc[train_idx]['var1'].unique()
    shape_v1 = len(uniq_v1)
    uniq_v2 = X.iloc[train_idx]['var2'].unique()
    shape_v2 = len(uniq_v2)
    ...
    obs = X.iloc[train_idx]['observation']
    shape_obs = len(train_idx)

    with pymc3.Model() as model:
        alpha = pm.Normal('alpha', mu = 0, sd = 1, shape = shape_v1)
        beta = pm.Normal('beta', mu = 0, sd = 1, shape = shape_v2)
        p = generic_function_of_alpha_and_beta
        posterior = pymc3.Bernoulli('posterior', p = p, observed = obs)

    solution = pymc3.find_MAP(model = model)
```
But as I understand it now, theano keeps copies of alpha and beta in its own memory space, and essentially allocates new copies of the array every time I define the model in the loop.  I've tried something along the lines of:
```
X = pandas.DataFrame()
shapes = theano.shared(np.ones(num_vars_to_fit), borrow = True))
obs = theano.shared(np.ones(1), borrow = True)
my_model = pymc3.Model()

with my_model:
    alpha = pm.Normal('alpha', mu = 0, sd = 1, shape = shapes.get_value()[0])
    beta = pm.Normal('beta', mu = 0, sd = 1, shape = shapes.get_value()[1])
    p = generic_function_of_alpha_and_beta
     posterior = pymc3.Bernoulli('posterior', p = p, observed = obs.get_value())

Fold = sklearn.model_selection.KFold(n_splits = n_splits):
for train_idx, test_idx in Fold.split(df):
    uniq_v1 = X.iloc[train_idx]['var1'].unique()
    shape_v1 = len(uniq_v1)
    uniq_v2 = X.iloc[train_idx]['var2'].unique()
    shape_v2 = len(uniq_v2)
    ...
    obs.set_value(X.iloc[train_idx]['observation'])
    shapes.set_value(numpy.array([shape_v1, shape_v2]))
    solution = pymc3.find_MAP(model = my_model)
```
But this doesn't seem to want to modify the shape parameter to alpha or beta, so find_MAP returns arrays that are the wrong size.  I'm just trying to get my head around how PyMC3 uses theano to roll the model definitions into C code snippets for efficiency.

**Comment by hvasbath** [comment]
@aaronjmaxwell in your case having shared variables doesnt change anything. The shape of input variables to a function must not change! If you can construct your model in a way that the shape somehow changes an intermediate variable within your graph (not the beginning or the end) then you can build a model that does not need to be recompiled. But in the way you have it now it always has to be recompiled as the shape of your input (start of graph) changes.

But as I mentioned above you can wrap everythin in the way that it hasnt to be like that. For example if you use as an input only a scalar and its value determines the size of your intermediate graph nodes, you can have it flexible. I guess my explanation is not enough now to fully understand, but I recommend you to do the theano tutorials. 

The with model block contains the theano graph which is then compiled by pymc3 to a function that only returns the likelihood. For further details the source code is open source, wasnt it? :)

**Comment by aaronjmaxwell** [comment]
Ah, I see @hvasbath.  Okay, so the theano graph has to be recompiled when the shape of the distribution changes.  But under my original implementation, where I redefine the model with it's new shape on every iteration, my memory usage jumped every time.  I looked through some of the theano tutorials and I couldn't find anything that specifically told me how to free up the memory used by the compiled graph.

**Comment by nouiz** [comment]
I didn't re-read the full thread, so I could me missing something. But
Theano allow to have the shapes of the inputs to changes, it the user graph
accept it. But sometimes the user graph add restriction to what the input
shapes must be.

**Comment by aaronjmaxwell** [comment]
@nouiz my problem is not with shared variables or changing the shape.  My problem is that if I have to compile the same model every time the shape of the input changes, python's memory usage jumps dramatically.  This is what I was trying to explain to @hvasbath in my contributions to this thread.  I've tried python's del statements combined with gc.collect() calls numerous times, and for whatever reason pymc3 (or theano) does not give the memory used by the previous model compilation back to the python memory manager.  So my OS memory eventually gets all used up.  To give you an idea, if you take the code snippet from the pymc3 tutorial page on linear regression, and place that code in a loop, the memory usage by python increases even though, in principle, everything is being redefined.

**Comment by nouiz** [comment]
If you don't enable Theano profiler, theano should be able to completly
delete compiled Theano functions I don't know if pymc3 is keeping global
variable that would keep reference to Theano?

What I was tring to tell, is that for Theano, you don't need to build new
graph or compile new Theano function if the input shape changes except if
that changes the computation. But maybe pymc3 add this restriction (could
be due to how they use compute_test_value)

**Comment by aaronjmaxwell** [comment]
If anyone is wondering how I solved my problem, see #1959
